File: C++/chess_demo.cpp
#include "common.hpp"
#include "board.hpp"
#include "spot.hpp"
#include "piece.hpp"
#include "piece_pawn.hpp"
#include "piece_bishop.hpp"
#include "piece_rook.hpp"
#include "player.hpp"
#include "game.hpp"

//Command Pattern to store history, undo, redo
//Different chess games according to timings
//Rankings
//Player Matching Strategy
//Online Tournaments - types of tournaments - Factory
//Game Controller
//Login, Payment, Notifications, Logging - general requirements of systems

int main() {

    Player* player1 = new Player("keerti", true);
    Player* player2 = new Player("amit", false);

    Game* game = new Game(player1, player2);
    Player* currPlayer = player1;

    while(1) {
        string start, end;
        cout<<"Please enter start spot"<<endl;
        cin>>start;
        if(start == "q") {
            cout<<"GAME OVER"<<endl;
            cout<<"-----------------------------------------------"<<endl;
            break;
        }
        else if(start == "u") {
            game->undo();
            continue;
        }
        cout<<"Please enter end spot"<<endl;
        cin>>end;
        if(end == "q") {
            cout<<"GAME OVER"<<endl;
            cout<<"-----------------------------------------------"<<endl;
            break;
        }
        if(game->processMove(player1, start, end))
        {
            if(currPlayer == player1)
                currPlayer = player2;
            else currPlayer = player1;
            cout<<"It is "<<currPlayer->getName()<<"'s turn now"<<endl;
        }
    }

    return 0;
}

File: JAVA/ChessDemo.java
import java.util.*;

public class ChessDemo {
    public static void main(String[] args) {
        Player player1 = new Player("keerti", true);
        Player player2 = new Player("amit", false);

        Game game = new Game(player1, player2);
        Player currPlayer = player1;

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Please enter start spot:");
            String start = scanner.next();
            if (start.equals("q")) {
                System.out.println("GAME OVER");
                System.out.println("-----------------------------------------------");
                break;
            } else if (start.equals("u")) {
                game.undo();
                continue;
            }

            System.out.println("Please enter end spot:");
            String end = scanner.next();
            if (end.equals("q")) {
                System.out.println("GAME OVER");
                System.out.println("-----------------------------------------------");
                break;
            }

            if (game.processMove(player1, start, end)) {
                if (currPlayer == player1)
                    currPlayer = player2;
                else
                    currPlayer = player1;
                System.out.println("It is " + currPlayer.getName() + "'s turn now");
            }
        }
    }
}
